execute/executor.go

```diff
@@ -119,77 +117,106 @@ type triggeringSpec interface {
 	TriggerSpec() flux.TriggerSpec
 }
 
-func (es *executionState) createNode(ctx context.Context, pr *plan.Procedure, nodes map[plan.ProcedureID]Node) (Node, error) {
-	// Check if we already created this node
-	if n, ok := nodes[pr.ID]; ok {
-		return n, nil
-	}
+// createExecutionNodeVisitor visits each node in a physical query plan
+// and creates a node responsible for executing that physical operation.
+type createExecutionNodeVisitor struct {
+	ctx   context.Context
+	es    *executionState
+	nodes map[plan.PlanNode]Node
+}
 
-	// Add explicit stream context if bounds are set on this node
-	var streamContext streamContext
-	if pr.Bounds != nil {
-		streamContext.bounds = &Bounds{
-			Start: pr.Bounds.Start,
-			Stop:  pr.Bounds.Stop,
-		}
-	}
+// Visit creates the node that will execute a particular plan node
+func (v *createExecutionNodeVisitor) Visit(node plan.PlanNode) error {
+	spec := node.ProcedureSpec()
+	kind := spec.Kind()
+	id := plan.ProcedureIDFromNodeID(node.ID())
 
 	// Build execution context
 	ec := executionContext{
-		ctx:           ctx,
-		es:            es,
-		streamContext: streamContext,
+		ctx:     v.ctx,
+		es:      v.es,
+		parents: make([]DatasetID, len(node.Predecessors())),
 	}
 
-	if len(pr.Parents) > 0 {
-		ec.parents = make([]DatasetID, len(pr.Parents))
-		for i, parentID := range pr.Parents {
-			ec.parents[i] = DatasetID(parentID)
-		}
+	for i, pred := range node.Predecessors() {
+		id := plan.ProcedureIDFromNodeID(pred.ID())
+		ec.parents[i] = DatasetID(id)
 	}
 
-	// If source create source
-	if createS, ok := procedureToSource[pr.Spec.Kind()]; ok {
-		s, err := createS(pr.Spec, DatasetID(pr.ID), ec)
+	// TODO what to do about stream context?
+
+	switch {
+	case len(node.Successors()) == 0:
+		// If node is a root, create a result.
+		// No need to add transports to predecessors as results pull data from downstream.
+		yield, ok := spec.(plan.YieldProcedureSpec)
+
+		if !ok {
+			return fmt.Errorf("expected yield, instead got %T", spec)
+		}
+
+		numPreds := len(node.Predecessors())
+
+		if numPreds != 1 {
```

Comment:
There can be only one predecessor for a yield?  I wonder why.

Reply:
Yes, you're right.  The user can put in a `union` if they want to do this.  I was confused.